fendermint/eth/hardhat/src/lib.rs

```diff
@@ -0,0 +1,386 @@
+
+/// Fully Qualified Name of a contract, e.g. `"src/lib/SubnetIDHelper.sol:SubnetIDHelper"`.
+pub type FQN = String;
+
+/// Dependency tree for libraries.
+///
+/// Using a [BTreeMap] for deterministic ordering.
+type DependencyTree<T> = BTreeMap<T, HashSet<T>>;
+
+/// Utility to link bytecode from Hardhat build artifacts.
+#[derive(Clone, Debug)]
+pub struct Hardhat {
+    /// Directory with Hardhat build artifacts, the full-fat JSON files
+    /// that contain ABI, bytecode, link references, etc.
+    contracts_dir: PathBuf,
+}
+
+impl Hardhat {
+    pub fn new(contracts_dir: PathBuf) -> Self {
+        Self { contracts_dir }
+    }
+
+    /// Fully qualified name of a source and contract.
+    pub fn fqn(&self, contract_source: &Path, contract_name: &str) -> String {
+        format!("{}:{}", contract_source.to_string_lossy(), contract_name)
+    }
+
+    /// Read the bytecode of the contract and replace all links in it with library addresses,
+    /// similar to how the [hardhat-ethers](https://github.com/NomicFoundation/hardhat/blob/7cc06ab222be8db43265664c68416fdae3030418/packages/hardhat-ethers/src/internal/helpers.ts#L165C42-L165C42)
+    /// plugin does it.
+    ///
+    /// The contract source is expected to be the logical path to a Solidity contract,
+    /// including the extension, ie. a [ContractSource].
+    pub fn bytecode(
+        &self,
+        contract_src: impl AsRef<Path>,
+        contract_name: &str,
+        libraries: &HashMap<FQN, et::Address>,
+    ) -> anyhow::Result<Vec<u8>> {
+        let artifact = self.artifact(contract_src.as_ref(), contract_name)?;
+
+        // Get the bytecode which is in hex format with placeholders for library references.
+        let mut bytecode = artifact.bytecode.object.clone();
+
+        // Replace all library references with their address.
+        // Here we differ slightly from the TypeScript version in that we don't return an error
+        // for entries in the library address map that we end up not needing, so we can afford
+        // to know less about which contract needs which exact references when we call them,
+        for (lib_src, lib_name) in artifact.libraries_needed() {
+            // References can be given with Fully Qualified Name, or just the contract name,
+            // but they must be unique and unambiguous.
+            let fqn = self.fqn(&lib_src, &lib_name);
+
+            let lib_addr = match (libraries.get(&fqn), libraries.get(&lib_name)) {
+                (None, None) => {
+                    bail!("failed to resolve library: {fqn}")
+                }
+                (Some(_), Some(_)) => bail!("ambiguous library: {fqn}"),
+                (Some(addr), None) => addr,
+                (None, Some(addr)) => addr,
+            };
+
+            let lib_addr = hex::encode(lib_addr.0);
+
+            for pos in artifact.library_positions(&lib_src, &lib_name) {
+                let start = 2 + pos.start * 2;
```

Comment:
In the docstring I included a link to the TypeScript plugin, albeit not a very precise one (and an accidental highlight). 

The logic was taken from [here](https://github.com/NomicFoundation/hardhat/blob/7cc06ab222be8db43265664c68416fdae3030418/packages/hardhat-ethers/src/internal/helpers.ts#L422-L436). 

My deduction if I had to explain it would be that the position `start` and `length` are in terms of actual bytes, indices in the byte vector. Since we are dealing with a hexadecimal string here, which has the `0x` prefix, the transformation has to have the `2 + ` in the beginning to skip the prefix, then `start * 2` because every byte is 2 characters in hex, and then `length * 2` for the same reason. 

As it happens the `__$a0fe38a9a81205afccc10810b09a4d3f45$__` and similar placeholders are exactly 40 characters, including the `__$` and `$__` so it all works out.